File: homework4/design_pattern/kotlin/factory.kt
// Define the Animal interface
interface Animal {
    fun speak(): String
}

// Implement the Dog class
class Dog : Animal {
    override fun speak(): String {
        return "Woof!"
    }
}

// Implement the Cat class
class Cat : Animal {
    override fun speak(): String {
        return "Meow!"
    }
}

// Define the AnimalFactory class
class AnimalFactory {
    fun createAnimal(animalType: String): Animal? {
        return when (animalType.lowercase()) {
            "dog" -> Dog()
            "cat" -> Cat()
            else -> null
        }
    }
}

// Example usage
fun main() {
    val factory = AnimalFactory()

    val dog = factory.createAnimal("dog")
    println(dog?.speak()) // Output: Woof!

    val cat = factory.createAnimal("cat")
    println(cat?.speak()) // Output: Meow!
}

File: homework4/design_pattern/csharp/factory.cs
using System;

namespace DesignPattern.Factory
{
    // Abstract base class
    public abstract class Animal
    {
        public abstract string Speak();
    }

    // Concrete class Dog
    public class Dog : Animal
    {
        public override string Speak()
        {
            return "Woof!";
        }
    }

    // Concrete class Cat
    public class Cat : Animal
    {
        public override string Speak()
        {
            return "Meow!";
        }
    }

    // Factory class
    public class AnimalFactory
    {
        public Animal CreateAnimal(string animalType)
        {
            if (animalType == "dog")
            {
                return new Dog();
            }
            else if (animalType == "cat")
            {
                return new Cat();
            }
            else
            {
                throw new ArgumentException("Invalid animal type");
            }
        }
    }

    // Example usage
    class Program
    {
        static void Main(string[] args)
        {
            AnimalFactory factory = new AnimalFactory();

            Animal dog = factory.CreateAnimal("dog");
            Console.WriteLine(dog.Speak()); // Output: Woof!

            Animal cat = factory.CreateAnimal("cat");
            Console.WriteLine(cat.Speak()); // Output: Meow!
        }
    }
}